pilot/platform/cloudfoundry/servicediscovery.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudfoundry
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+
+	copilotapi "code.cloudfoundry.org/copilot/api"
+
+	"istio.io/istio/pilot/model"
+)
+
+// CopilotClient defines a local interface for interacting with Cloud Foundry Copilot
+type CopilotClient interface {
+	copilotapi.IstioCopilotClient
+}
+
+// AppPort is the container-side port on which Cloud Foundry Diego applications listen
+const AppPort = 8080
+
+// ServiceDiscovery implements the model.ServiceDiscovery interface for Cloud Foundry
+type ServiceDiscovery struct {
+	Client CopilotClient
+}
+
+// Services implements a service catalog operation
+func (sd *ServiceDiscovery) Services() ([]*model.Service, error) {
+	resp, err := sd.Client.Routes(context.Background(), new(copilotapi.RoutesRequest))
+	if err != nil {
+		return nil, fmt.Errorf("getting services: %s", err)
+	}
+	services := make([]*model.Service, 0, len(resp.GetBackends()))
+
+	for hostname := range resp.Backends {
+		services = append(services, newService(hostname))
+	}
+
+	return services, nil
+}
+
+// GetService implements a service catalog operation
+func (sd *ServiceDiscovery) GetService(hostname string) (*model.Service, error) {
+	services, err := sd.Services()
+	if err != nil {
+		return nil, err
+	}
+	for _, svc := range services {
+		if svc.Hostname == hostname {
+			return svc, nil
+		}
+	}
+	return nil, fmt.Errorf("no service exists with name '%s'", hostname)
```

Comment:
I'm not sure this should return `error` in the case no services are found.  I think `error` should be reserved for problems communicating with the service registry or a bad hostname was provided.  I believe this is how `error` is handled in the consul and eureka adapters

Reply:
Good catch, thank you.  Fixed in https://github.com/istio/istio/pull/2022/commits/4419481abd423a58943f42b04ae81c81549653d5